### PR TITLE
Disallow Unsized Enums

### LINKED ISSUE #16812: ICE: Unexpected type returned from struct_tail
```rust
enum Foo {
    Bar(int, [int]),
}

fn main() {
    // Either of these lines can cause the ICE
    let _x: &(int, [int]);
    let _y: &Foo;
}
```

Backtrace:

```
   1:     0x7f2a127e8580 - rt::backtrace::imp::write::h0180e7ded76a3ff2c0q
   2:     0x7f2a127eb770 - failure::on_fail::h3f169393c89b5f90Tlr
   3:     0x7f2a12fd24f0 - unwind::begin_unwind_inner::haeb9517b5061bc86Zie
   4:     0x7f2a133ad670 - unwind::begin_unwind::h17017358962417336117
   5:     0x7f2a13803a20 - middle::ty::unsized_part_of_type::h8eae6d239e9b7af3tnG
   6:     0x7f2a138037c0 - middle::trans::type_of::type_of::type_of_unsize_info::h67819f63aa2acfb5Xh9
   7:     0x7f2a1377b050 - middle::trans::type_of::type_of::h3c89a1738d384c45Ih9
   8:     0x7f2a1377b050 - middle::trans::type_of::type_of::h3c89a1738d384c45Ih9
   9:     0x7f2a137b0860 - middle::trans::base::alloc_ty::h5485998eeb61d023mTd
  10:     0x7f2a13859870 - middle::trans::_match::mk_binding_alloca::h11628361498234501597
  11:     0x7f2a1385bdf0 - middle::trans::_match::store_local::create_dummy_locals::closure.119283
  12:     0x7f2a13681560 - middle::pat_util::pat_bindings::closure.112016
  13:     0x7f2a0ea1ccb0 - ast_util::walk_pat::hc858153057fb5965oRB
  14:     0x7f2a138205f0 - middle::trans::_match::store_local::h9195017fe21b4124A2h
  15:     0x7f2a13772d10 - middle::trans::base::init_local::h5e8a493dc2ae05b1WFd
  16:     0x7f2a13772220 - middle::trans::controlflow::trans_stmt::hf26d5cee55a6056blhY
  17:     0x7f2a13773c40 - middle::trans::controlflow::trans_block::h004dea36138a75c8wmY
  18:     0x7f2a13829560 - middle::trans::base::trans_closure::h655acfc433ac13fevye
  19:     0x7f2a13764670 - middle::trans::base::trans_fn::hf0a444fa9ff43aebiKe
  20:     0x7f2a1375f9b0 - middle::trans::base::trans_item::hda4d4160cb2b58dfi2e
  21:     0x7f2a138341f0 - middle::trans::base::trans_crate::hff0f87c98e6356c8JWf
  22:     0x7f2a13c058d0 - driver::driver::phase_4_translate_to_llvm::h8223ee9d7ab4951cpzy
  23:     0x7f2a13bfd280 - driver::driver::compile_input::hfa4fd95c9e6a2ceeQby
  24:     0x7f2a13c8f0a0 - driver::run_compiler::h32bdccca6ed9465cJIB
  25:     0x7f2a13c8efb0 - driver::main_args::closure.138206
  26:     0x7f2a13ca1240 - task::TaskBuilder<S>::try_future::closure.139324
  27:     0x7f2a13ca1040 - task::TaskBuilder<S>::spawn_internal::closure.139301
  28:     0x7f2a146dcf10 - task::spawn_opts::closure.8369
  29:     0x7f2a1302af90 - rust_try_inner
  30:     0x7f2a1302af80 - rust_try
  31:     0x7f2a12fcfb20 - unwind::try::h03d8d1d4cb0de0c1f7d
  32:     0x7f2a12fcf8e0 - task::Task::run::hafab6bcab45e61e5zdd
  33:     0x7f2a146dcc70 - task::spawn_opts::closure.8315
  34:     0x7f2a12fd16e0 - thread::thread_start::hba588af1f803bb5blCd
  35:     0x7f2a1227f0c0 - start_thread
  36:     0x7f2a12c99359 - __clone
  37:                0x0 - <unknown>
```

This seems to occur when using a non-struct type that contains an unsized type. This is probably because [`unsized_part_of_type`](https://github.com/rust-lang/rust/blob/7932b719ec2b65acfa8c3e74aad29346d47ee992/src/librustc/middle/ty.rs#L2938-L2956) only handles `ty_struct` and the base unsized types. This should be extended to handle enums and tuples as well.

### PR DESCRIPTION
**Pull request by TimNN**
Fixes #16812.

This PR is a potential fix for #16812, an issue which is reported [again](https://github.com/rust-lang/rust/issues/36801) and [again](https://github.com/rust-lang/rust/issues/36975), with over a dozen duplicates by now.

This PR is mainly meant to promoted discussion about the issue and the correct way to fix it.

This is a [breaking-change] since the error is now reported during wfchecking, so that even the definition of a (potentially) unsized enum will cause an error (whereas it would previously cause an ICE at trans time if the enum was used in an unsized manner).

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Hmm. So this is *intended* to work -- but I agree that it'd be better to give a clean error. Probably this would require a crater run to get some idea.

**Comment by nikomatsakis** [comment]
cc @rust-lang/compiler @rust-lang/lang and @eddyb in particular =)

This PR proposes to remove the ability for an enum to be unsized. They were intended to support this, but it was never fuly implemented, and the compiler currently ICEs if you really try to do anything. There is some backwards compatbility risk here, but it's pretty minor: you have to have enums that are potentially unsized but not used as such, I think. I'll try to kick off a crater run to get more data.

I think I'm in favor of making them a hard error (to fix the ICEs) but opening an issue to implement the support properly.

**Comment by nrc** [comment]
How difficult would it be to fix unsized enums? Is there design work to be done there or does it just need the implementation finished off?

**Comment by eddyb** [comment]
@nrc I don't think anything about them was ever specified, and I'd expect a RFC or at least ammendments.

**Comment by nikomatsakis** [comment]
@eddyb 

> I don't think anything about them was ever specified, and I'd expect a RFC or at least ammendments.

Can you say a bit more here? Off the top of my head I'm not sure what complications you are imagining would require amending RFCs and the like.

**Comment by eddyb** [comment]
@nikomatsakis There are no defined semantics. Which cases are allowed? One maybe-unsized variant? Potentially all of them? Does the pointer metadata behave as an union tagged by the enum itself?
(Right now we have no unsized types that need the pointer to determine information like size/align)

I suppose `CoerceUnsized` is orthogonal so there's less to define than I initially thought.

If we can agree on something and we can implement it by reusing univariant functionality in most places (as opposed to needing a lot of special logic), then I'm okay with just a RFC-less FCP.

**Comment by nikomatsakis** [comment]
Started crater run from a29c49f5cca849cd5ac79b3cd70b934bb33cf4e6 to c5483ff3b9fb8b5fdb633707b3e38c66c83ddae6.

**Comment by nikomatsakis** [comment]
Crater results: https://gist.github.com/13d56401d4ceee692d4ad9374c8d3600

- 6540 crates tested: 4832 working / 1648 broken / **6 regressed** / 0 fixed / 54 unknown.

* [ ] [graph-io-gml-0.1.1](https://crates.io/crates/graph-io-gml) ([before](https://tools.taskcluster.net/task-inspector/#DHUEVffPQTua8xuyo7iFTg)) ([after](https://tools.taskcluster.net/task-inspector/#bfQpi-x0ReKmlOkxRvlZuQ))
* [ ] [lpsolve-sys-5.5.0](https://crates.io/crates/lpsolve-sys) ([before](https://tools.taskcluster.net/task-inspector/#CVj7r8rWRIS32l8MlZEc4Q)) ([after](https://tools.taskcluster.net/task-inspector/#nAEhI3-qR16WmPsFOPjmwg))
* [ ] [google-dfareporting2d4-1.0.0+20160803](https://crates.io/crates/google-dfareporting2d4) ([before](https://tools.taskcluster.net/task-inspector/#kla7_yrsTuGMubW1Fy5OjA)) ([after](https://tools.taskcluster.net/task-inspector/#9Kn-uFj_S32s1V9R-jOnoQ))
* [ ] [google-fusiontables2-1.0.0+20160526](https://crates.io/crates/google-fusiontables2) ([before](https://tools.taskcluster.net/task-inspector/#zIRvpQh3QFSjJYJL1zEIVA)) ([after](https://tools.taskcluster.net/task-inspector/#E32EdXFlS_6_4wx_1-464Q))
* [ ] [google-logging2_beta1-cli-1.0.0+20160322](https://crates.io/crates/google-logging2_beta1-cli) ([before](https://tools.taskcluster.net/task-inspector/#uScson_DRbOhX_ebrLVbKw)) ([after](https://tools.taskcluster.net/task-inspector/#eYHIA62oSyyObXoqZEGtLg))
* [ ] [drawille-0.2.3](https://crates.io/crates/drawille) ([before](https://tools.taskcluster.net/task-inspector/#tygT2ASnRqep1QurKsGE9w)) ([after](https://tools.taskcluster.net/task-inspector/#GbhxRPEhR3OScovVwjxzWw))

**Comment by eddyb** [comment]
@nikomatsakis The google crates are auto-generated, ouch. I'd rather just make it work then :disappointed:.

**Comment by nikomatsakis** [comment]
Appears to be no regressions. I'm inclined to think we should just make this an error for now. I don't know that I think that an RFC is necessarily required to implement these changes, but it couldn't hurt in any case -- and ICEing is no good.

@rfcbot fcp merge

**Comment by nikomatsakis** [comment]
@eddyb I think it's all false, no? Though we could try (and perhaps should) testing those crates by hand.

**Comment by TimNN** [comment]
Great! Then I'll see about updating / adding tests in the next few days.

**Comment by TimNN** [comment]
I could successfully build all listed "regressions" with a local stage 1 of c5483ff3b9fb8b5fdb633707b3e38c66c83ddae6.

**Comment by TimNN** [comment]
I have adapted and added the test cases.

File: src/test/compile-fail/issue-17025.rs [hunk 1]
```diff
@@ -8,6 +8,8 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// ignore-test the unsized enum no longer compiles
```
Comment (TimNN) [inline src/test/compile-fail/issue-17025.rs]: I have ignored this test for now, but it has effectively become useless and could be deleted but I'm unsure what the policy is in this case.

**Comment by withoutboats** [comment]
@rfcbot reviewed

Seems like making this an error is the correct thing to do for now.

**Comment by nikomatsakis** [comment]
thanks tim

**Pull request merged**